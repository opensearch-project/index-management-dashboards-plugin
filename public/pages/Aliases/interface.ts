/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
export interface IAlias {
  alias: string;
  index: string;
  filter: string;
  "routing.index": string;
  "routing.search": string;
  is_write_index: string;
  indexArray: string[];
}
