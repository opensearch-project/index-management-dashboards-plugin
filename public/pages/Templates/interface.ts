/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import { TemplateItemRemote } from "../../../models/interfaces";

export interface ITemplate {
  name: string;
  index_patterns: string;
  order: number;
  version?: string;
  composed_of: string;
  templateDetail?: TemplateItemRemote;
}
