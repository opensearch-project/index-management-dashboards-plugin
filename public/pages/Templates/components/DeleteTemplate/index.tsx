/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import React, { useState } from "react";
import { EuiButtonIcon, EuiToolTip } from "@elastic/eui";
import DeleteIndexModal from "../../containers/DeleteTemplatesModal";
import { ITemplate } from "../../interface";

export interface DeleteTemplateProps {
  selectedItems: ITemplate[];
  onDelete: () => void;
}

export default function DeleteTemplate(props: DeleteTemplateProps) {
  const { selectedItems, onDelete } = props;
  const [deleteIndexModalVisible, setDeleteIndexModalVisible] = useState(false);

  const onDeleteIndexModalClose = () => {
    setDeleteIndexModalVisible(false);
  };

  return (
    <>
      <EuiToolTip content="Delete template">
        <EuiButtonIcon
          aria-label="Delete template"
          color="danger"
          iconType="trash"
          onClick={() => setDeleteIndexModalVisible(true)}
          className="icon-hover-danger"
        />
      </EuiToolTip>
      <DeleteIndexModal
        selectedItems={selectedItems.map((item) => item.name)}
        visible={deleteIndexModalVisible}
        onClose={onDeleteIndexModalClose}
        onDelete={() => {
          onDeleteIndexModalClose();
          onDelete();
        }}
      />
    </>
  );
}
