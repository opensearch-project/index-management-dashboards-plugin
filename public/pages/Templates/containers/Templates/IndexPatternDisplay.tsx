/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useState } from "react";
import { Criteria, EuiBasicTable, EuiFlyout, EuiFlyoutBody, EuiFlyoutHeader, EuiLink, EuiText } from "@elastic/eui";

export default function IndexPatternDisplay(props: { indexPatterns: string[]; templateName: string }) {
  const [hide, setHide] = useState(true);
  const [tableParams, setTableParams] = useState<Criteria<{ pattern: string }>>({});
  const { index, size } = tableParams.page || {
    index: 0,
    size: 10,
  };

  return (
    <div>
      <span>{props.indexPatterns.slice(0, 3).join(", ")}</span>
      {props.indexPatterns.length <= 3 ? null : (
        <EuiLink style={{ marginLeft: 8 }} data-test-subj={`${props.indexPatterns.length - 3} more`} onClick={() => setHide(!hide)}>
          {props.indexPatterns.length - 3} more
        </EuiLink>
      )}
      {hide ? null : (
        <EuiFlyout onClose={() => setHide(!hide)}>
          <EuiFlyoutHeader hasBorder>
            <EuiText size="m">
              <h2 title={`Index patterns in ${props.templateName} (${props.indexPatterns.length})`}>
                Index patterns in in {props.templateName} ({props.indexPatterns.length})
              </h2>
            </EuiText>
          </EuiFlyoutHeader>
          <EuiFlyoutBody>
            <EuiBasicTable
              data-test-subj="indexPatternsTable"
              columns={[
                {
                  name: "Index pattern",
                  field: "pattern",
                },
              ]}
              items={props.indexPatterns.slice(index * size, (index + 1) * size).map((pattern) => ({ pattern }))}
              onChange={setTableParams}
              pagination={{
                pageIndex: index,
                pageSize: size,
                totalItemCount: props.indexPatterns.length,
              }}
            />
          </EuiFlyoutBody>
        </EuiFlyout>
      )}
    </div>
  );
}
