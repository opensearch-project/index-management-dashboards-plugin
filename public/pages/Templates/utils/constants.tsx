/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import { SortDirection } from "../../../utils/constants";
import { ITemplate } from "../interface";

export const DEFAULT_PAGE_SIZE_OPTIONS = [5, 10, 20, 50];

export const DEFAULT_QUERY_PARAMS = {
  from: "0",
  size: "20",
  search: "",
  sortField: "name" as keyof ITemplate,
  sortDirection: SortDirection.DESC,
  dataSourceId: "",
};
