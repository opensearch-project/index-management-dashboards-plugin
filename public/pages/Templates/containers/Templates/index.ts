/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import Templates from "./Templates";

export default Templates;
