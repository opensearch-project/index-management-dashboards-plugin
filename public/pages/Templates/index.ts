/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import Templates from "./containers/Templates";

export default Templates;
