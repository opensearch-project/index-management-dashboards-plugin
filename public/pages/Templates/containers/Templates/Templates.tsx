/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useContext } from "react";
import _, { debounce, isEqual } from "lodash";
import { Link, RouteComponentProps } from "react-router-dom";
import queryString from "query-string";
import {
  EuiHorizontalRule,
  EuiBasicTable,
  Criteria,
  EuiTableSortingType,
  Direction,
  Pagination,
  EuiTableSelectionType,
  EuiSmallButton,
  EuiLink,
  EuiTitle,
  EuiCompressedFormRow,
  EuiEmptyPrompt,
  EuiText,
  EuiToolTip,
  EuiSmallButtonIcon,
  EuiPanel,
} from "@elastic/eui";
import { ContentPanel, ContentPanelActions } from "../../../../components/ContentPanel";
import { DEFAULT_PAGE_SIZE_OPTIONS, DEFAULT_QUERY_PARAMS } from "../../utils/constants";
import CommonService from "../../../../services/CommonService";
import { ITemplate } from "../../interface";
import { BREADCRUMBS, PLUGIN_NAME, ROUTES } from "../../../../utils/constants";
import { CoreServicesContext } from "../../../../components/core_services";
import { ServicesContext } from "../../../../services";
import IndexControls, { SearchControlsProps } from "../../components/IndexControls";
import TemplatesActions from "../TemplatesActions";
import { CoreStart } from "opensearch-dashboards/public";
import { TemplateItemRemote } from "../../../../../models/interfaces";
import { TemplateConvert } from "../../../CreateIndexTemplate/components/TemplateType";
import AssociatedComponentsModal from "../AssociatedComponentsModal";
import DeleteTemplate from "../../components/DeleteTemplate";
import IndexPatternDisplay from "./IndexPatternDisplay";
import { DataSourceMenuContext, DataSourceMenuProperties } from "../../../../services/DataSourceMenuContext";
import MDSEnabledComponent from "../../../../components/MDSEnabledComponent";
import { getApplication, getNavigationUI, getUISettings } from "../../../../services/Services";
import { TopNavControlButtonData, TopNavControlDescriptionData, TopNavControlLinkData } from "src/plugins/navigation/public";

interface TemplatesProps extends RouteComponentProps, DataSourceMenuProperties {
  commonService: CommonService;
}

type TemplatesState = {
  totalTemplates: number;
  from: string;
  size: string;
  sortField: keyof ITemplate;
  sortDirection: Direction;
  selectedItems: ITemplate[];
  templates: ITemplate[];
  loading: boolean;
  useUpdatedUX: boolean;
} & SearchControlsProps["value"] &
  DataSourceMenuProperties;

const defaultFilter = {
  search: DEFAULT_QUERY_PARAMS.search,
};

class Templates extends MDSEnabledComponent<TemplatesProps, TemplatesState> {
  static contextType = CoreServicesContext;
  constructor(props: TemplatesProps) {
    super(props);
    const {
      from = DEFAULT_QUERY_PARAMS.from,
      size = DEFAULT_QUERY_PARAMS.size,
      search = DEFAULT_QUERY_PARAMS.search,
      sortField = DEFAULT_QUERY_PARAMS.sortField,
      sortDirection = DEFAULT_QUERY_PARAMS.sortDirection,
    } = queryString.parse(props.history.location.search) as {
      from: string;
      size: string;
      search: string;
      sortField: keyof ITemplate;
      sortDirection: Direction;
    };

    const uiSettings = getUISettings();
    const useUpdatedUX = uiSettings.get("home:useNewHomePage");

    this.state = {
      ...defaultFilter,
      ...this.state,
      totalTemplates: 0,
      from,
      size,
      search,
      sortField,
      sortDirection,
      selectedItems: [],
      templates: [],
      loading: false,
      useUpdatedUX: useUpdatedUX,
    };

    this.getTemplates = debounce(this.getTemplates, 500, { leading: true });
  }

  async componentDidUpdate(prevProps: TemplatesProps, prevState: TemplatesState) {
    const prevQuery = this.getQueryState(prevState);
    const currQuery = this.getQueryState(this.state);
    if (!_.isEqual(prevQuery, currQuery)) {
      await this.getTemplates();
    }
  }
  componentDidMount() {
    const breadCrumbs = this.state.useUpdatedUX ? [BREADCRUMBS.NEW_TEMPLATES] : [BREADCRUMBS.INDEX_MANAGEMENT, BREADCRUMBS.TEMPLATES];
    this.context.chrome.setBreadcrumbs(breadCrumbs);
    this.getTemplates();
  }

  getQueryState = (state: TemplatesState) => {
    return Object.keys(DEFAULT_QUERY_PARAMS).reduce((total, key) => {
      return {
        ...total,
        [key]: state[key as keyof typeof DEFAULT_QUERY_PARAMS],
      };
    }, {} as TemplatesState);
  };

  getTemplates = async (): Promise<void> => {
    this.setState({ loading: true });
    const { from, size } = this.state;
    const fromNumber = Number(from);
    const sizeNumber = Number(size);
    const { history, commonService } = this.props;
    const queryObject = this.getQueryState(this.state);
    const queryParamsString = queryString.stringify(queryObject);
    history.replace({ ...this.props.location, search: queryParamsString });

    const payload: any = {
      format: "json",
      name: `*${queryObject.search}*`,
      s: `${queryObject.sortField}:${queryObject.sortDirection}`,
    };

    let allTemplatesDetail: TemplateItemRemote[] = [];

    const allTemplatesResponse = await commonService.apiCaller<{
      index_templates?: {
        name: string;
        index_template: TemplateItemRemote;
      }[];
    }>({
      endpoint: "transport.request",
      data: {
        method: "GET",
        path: "/_index_template/*",
      },
    });

    if (!allTemplatesResponse.ok) {
      allTemplatesDetail = [];
    } else {
      allTemplatesDetail =
        allTemplatesResponse.response.index_templates?.map((item) => ({
          ...item.index_template,
          name: item.name,
        })) || [];
    }

    const getTemplatesResponse = await commonService.apiCaller<ITemplate[]>({
      endpoint: "cat.templates",
      data: payload,
    });

    if (getTemplatesResponse.ok) {
      // enhance the catResponse with template detail
      const response: ITemplate[] = getTemplatesResponse.response
        .filter((item) => item.composed_of)
        .map((item) => ({
          ...item,
          templateDetail: allTemplatesDetail.find((detailItem) => detailItem.name === item.name),
        }));
      const totalTemplates = response.length;
      const payload = {
        templates: response.slice(fromNumber * sizeNumber, (fromNumber + 1) * sizeNumber),
        totalTemplates,
        selectedItems: this.state.selectedItems
          .map((item) => response.find((remoteItem) => remoteItem.name === item.name))
          .filter((item) => item),
      } as TemplatesState;
      this.setState(payload);
    } else {
      this.context.notifications.toasts.addDanger(getTemplatesResponse.error);
    }

    // Avoiding flicker by showing/hiding the "Data stream" column only after the results are loaded.
    this.setState({ loading: false });
  };

  onTableChange = ({ page: tablePage, sort }: Criteria<ITemplate>): void => {
    const { index: page, size } = tablePage || {};
    const { field: sortField, direction: sortDirection } = sort || {};
    this.setState(
      {
        from: "" + page,
        size: "" + size,
        sortField: sortField || DEFAULT_QUERY_PARAMS.sortField,
        sortDirection: sortDirection as Direction,
      },
      () => this.getTemplates()
    );
  };

  onSelectionChange = (selectedItems: ITemplate[]): void => {
    this.setState({ selectedItems });
  };

  onSearchChange = (params: Parameters<SearchControlsProps["onSearchChange"]>[0]): void => {
    this.setState({ from: "0", ...params }, () => this.getTemplates());
  };

  render() {
    const { totalTemplates, from, size, sortField, sortDirection, templates } = this.state;

    const pagination: Pagination = {
      pageIndex: Number(from),
      pageSize: Number(size),
      pageSizeOptions: DEFAULT_PAGE_SIZE_OPTIONS,
      totalItemCount: Number(totalTemplates),
    };

    const sorting: EuiTableSortingType<ITemplate> = {
      sort: {
        direction: sortDirection,
        field: sortField,
      },
    };

    const selection: EuiTableSelectionType<ITemplate> = {
      onSelectionChange: this.onSelectionChange,
    };

    const commonTable = () => {
      return (
        <EuiBasicTable
          className="ISM-templates-table"
          data-test-subj="templatesTable"
          loading={this.state.loading}
          columns={[
            {
              field: "name",
              name: "Template name",
              sortable: true,
              render: (value: string) => {
                return (
                  <Link to={`${ROUTES.CREATE_TEMPLATE}/${value}`}>
                    <EuiLink>{value}</EuiLink>
                  </Link>
                );
              },
            },
            {
              field: "templateType",
              name: "Template type",
              render: (value: string, record) => {
                return TemplateConvert({
                  value: record.templateDetail?.data_stream,
                });
              },
            },
            {
              field: "index_patterns",
              name: "Index patterns",
              sortable: true,
              render: (value: string[], record) => {
                return <IndexPatternDisplay indexPatterns={record.templateDetail?.index_patterns || []} templateName={record.name} />;
              },
            },
            {
              field: "order",
              name: "Priority",
              sortable: true,
              align: "right",
            },
            {
              field: "composed_of",
              name: "Associated component templates",
              align: "right",
              render: (value: string, record) => record.templateDetail?.composed_of?.length || 0,
            },
            {
              field: "actions",
              name: "Actions",
              align: "right",
              actions: [
                {
                  render: (record: ITemplate) => (
                    <AssociatedComponentsModal
                      template={record}
                      onUnlink={() => this.getTemplates()}
                      renderProps={({ setVisible }) => (
                        <EuiToolTip content="View associated index templates">
                          <EuiSmallButtonIcon
                            aria-label="View associated index templates"
                            iconType="kqlSelector"
                            onClick={() => setVisible(true)}
                            className="icon-hover-info"
                          />
                        </EuiToolTip>
                      )}
                    />
                  ),
                },
                {
                  render: (record: ITemplate) => <DeleteTemplate selectedItems={[record]} onDelete={this.getTemplates} />,
                },
              ],
            },
          ]}
          isSelectable={true}
          itemId="name"
          items={templates}
          onChange={this.onTableChange}
          pagination={pagination}
          selection={selection}
          sorting={sorting}
          noItemsMessage={
            isEqual(
              {
                search: this.state.search,
              },
              defaultFilter
            ) ? (
              <EuiEmptyPrompt
                body={
                  <EuiText size="s">
                    <p>You have no templates.</p>
                  </EuiText>
                }
                actions={[
                  <EuiSmallButton
                    fill
                    onClick={() => {
                      this.props.history.push(ROUTES.CREATE_TEMPLATE);
                    }}
                    iconType="plus"
                  >
                    Create template
                  </EuiSmallButton>,
                ]}
              />
            ) : (
              <EuiEmptyPrompt
                body={
                  <EuiText size="s">
                    <p>There are no templates matching your applied filters. Reset your filters to view your templates.</p>
                  </EuiText>
                }
                actions={[
                  <EuiSmallButton
                    fill
                    onClick={() => {
                      this.setState(defaultFilter, () => {
                        this.getTemplates();
                      });
                    }}
                  >
                    Reset filters
                  </EuiSmallButton>,
                ]}
              />
            )
          }
        />
      );
    };

    const { HeaderControl } = getNavigationUI();
    const { setAppRightControls } = getApplication();
    const { setAppDescriptionControls } = getApplication();

    const description = [
      {
        description: "Index templates let you initialize new indexes or data streams with predefined mappings and settings.",
        links: {
          label: "Learn more",
          href: (this.context as CoreStart).docLinks.links.opensearch.indexTemplates.base,
          iconType: "popout",
          iconSide: "right",
          controlType: "link",
          target: "_blank",
          flush: "both",
        } as TopNavControlLinkData,
      } as TopNavControlDescriptionData,
    ];

    const onClickCreate = () => {
      this.props.history.push(ROUTES.CREATE_TEMPLATE);
    };

    return this.state.useUpdatedUX ? (
      <div style={{ padding: "0px" }}>
        <HeaderControl controls={description} setMountPoint={setAppDescriptionControls} />
        <HeaderControl
          setMountPoint={setAppRightControls}
          controls={[
            {
              id: "Create Index Template",
              label: "Create Index Template",
              fill: true,
              iconType: "plus",
              run: onClickCreate,
              testId: "createTemplateButton",
              controlType: "button",
              color: "primary",
            } as TopNavControlButtonData,
          ]}
        />
        <EuiPanel>
          <IndexControls
            value={{
              search: this.state.search,
            }}
            onSearchChange={this.onSearchChange}
            selectedItems={this.state.selectedItems}
            getTemplates={this.getTemplates}
            history={this.props.history}
          />

          {commonTable()}
        </EuiPanel>
      </div>
    ) : (
      <ContentPanel
        actions={
          <ContentPanelActions
            actions={[
              {
                text: "",
                children: (
                  <TemplatesActions selectedItems={this.state.selectedItems} onDelete={this.getTemplates} history={this.props.history} />
                ),
              },
              {
                text: "Create template",
                buttonProps: {
                  fill: true,
                  onClick: onClickCreate,
                  iconType: "plus",
                },
              },
            ]}
          />
        }
        bodyStyles={{ padding: "initial" }}
        title={
          <>
            <EuiTitle>
              <span>Templates</span>
            </EuiTitle>
            <EuiCompressedFormRow
              fullWidth
              helpText={
                <div>
                  Index templates let you initialize new indexes or data streams with predefined mappings and settings.{" "}
                  <EuiLink target="_blank" external href={(this.context as CoreStart).docLinks.links.opensearch.indexTemplates.base}>
                    Learn more
                  </EuiLink>
                </div>
              }
            >
              <></>
            </EuiCompressedFormRow>
          </>
        }
      >
        <IndexControls
          value={{
            search: this.state.search,
          }}
          onSearchChange={this.onSearchChange}
          selectedItems={this.state.selectedItems}
          getTemplates={this.getTemplates}
          history={this.props.history}
        />
        <EuiHorizontalRule margin="xs" />

        {commonTable()}
      </ContentPanel>
    );
  }
}

export default function TemplatesContainer(props: Omit<TemplatesProps, "commonService" | keyof DataSourceMenuProperties>) {
  const context = useContext(ServicesContext);
  const dataSourceMenuProps = useContext(DataSourceMenuContext);
  return <Templates {...props} commonService={context?.commonService as CommonService} {...dataSourceMenuProps} />;
}
