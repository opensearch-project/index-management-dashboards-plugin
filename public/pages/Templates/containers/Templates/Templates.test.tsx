/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render, waitFor } from "@testing-library/react";
import { Redirect, Route, Switch } from "react-router-dom";
import { HashRouter as Router } from "react-router-dom";
import { browserServicesMock, coreServicesMock } from "../../../../../test/mocks";
import Templates from "./index";
import { ServicesContext } from "../../../../services";
import { ROUTES } from "../../../../utils/constants";
import { CoreServicesContext } from "../../../../components/core_services";
import userEvent from "@testing-library/user-event";
import { ITemplate } from "../../interface";
import { getApplication, getNavigationUI, getUISettings } from "../../../../services/Services";

jest.mock("../../../../services/Services", () => ({
  ...jest.requireActual("../../../../services/Services"),
  getUISettings: jest.fn(),
  getApplication: jest.fn(),
  getNavigationUI: jest.fn(),
}));

beforeEach(() => {
  (getUISettings as jest.Mock).mockReturnValue({
    get: jest.fn().mockReturnValue(false), // or false, depending on your test case
  });
  (getApplication as jest.Mock).mockReturnValue({});

  (getNavigationUI as jest.Mock).mockReturnValue({});
});

function renderWithRouter() {
  return {
    ...render(
      <Router>
        <Switch>
          <Route
            path={ROUTES.TEMPLATES}
            render={(props) => (
              <CoreServicesContext.Provider value={coreServicesMock}>
                <ServicesContext.Provider value={browserServicesMock}>
                  <Templates {...props} />
                </ServicesContext.Provider>
              </CoreServicesContext.Provider>
            )}
          />
          <Redirect from="/" to={ROUTES.TEMPLATES} />
        </Switch>
      </Router>
    ),
  };
}

const testTemplateId = "test";

describe("<Templates /> spec", () => {
  beforeEach(() => {
    browserServicesMock.commonService.apiCaller = jest.fn(async (payload) => {
      if (payload.endpoint === "cat.templates") {
        return {
          ok: true,
          response: [
            {
              name: testTemplateId,
              index_patterns: "[1]",
              version: "",
              order: 1,
            },
          ] as ITemplate[],
        };
      }

      return {
        ok: true,
        response: {},
      };
    }) as any;
    window.location.hash = "/";
  });
  it("renders the component", async () => {
    const { container, getByTestId } = renderWithRouter();

    expect(container.firstChild).toMatchSnapshot();
    await waitFor(() => {
      expect(browserServicesMock.commonService.apiCaller).toBeCalledTimes(1);
    });
    userEvent.click(getByTestId("tableHeaderCell_name_0").querySelector("button") as Element);
    await waitFor(() => {
      expect(browserServicesMock.commonService.apiCaller).toBeCalledTimes(4);
      expect(browserServicesMock.commonService.apiCaller).toBeCalledWith({
        data: { format: "json", name: `**`, s: "name:asc" },
        endpoint: "cat.templates",
      });
    });
  });

  it("with some actions", async () => {
    const { getByPlaceholderText } = renderWithRouter();
    expect(browserServicesMock.commonService.apiCaller).toBeCalledTimes(1);
    userEvent.type(getByPlaceholderText("Search..."), `${testTemplateId}{enter}`);
    await waitFor(() => {
      expect(browserServicesMock.commonService.apiCaller).toBeCalledTimes(4);
      expect(browserServicesMock.commonService.apiCaller).toBeCalledWith({
        data: { format: "json", name: `*${testTemplateId}*`, s: "name:desc" },
        endpoint: "cat.templates",
      });
    });
  });
});
