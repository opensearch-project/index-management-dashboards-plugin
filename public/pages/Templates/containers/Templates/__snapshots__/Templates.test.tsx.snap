// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Templates /> spec renders the component 1`] = `
<div
  class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
  style="padding-left: 0px; padding-right: 0px;"
>
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsFlexStart euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
    style="padding: 0px 16px;"
  >
    <div
      class="euiFlexItem"
    >
      <span
        class="euiTitle euiTitle--medium"
      >
        Templates
      </span>
      <div
        class="euiFormRow euiFormRow--fullWidth euiFormRow--compressed"
        id="some_html_id-row"
      >
        <div
          class="euiFormRow__fieldWrapper"
        >
          <div
            class="euiFormHelpText euiFormRow__text"
            id="some_html_id-help-0"
          >
            <div>
              Index templates let you initialize new indexes or data streams with predefined mappings and settings.
               
              <a
                class="euiLink euiLink--primary"
                href="https://opensearch.org/docs/latest/opensearch/index-templates"
                rel="noopener noreferrer"
                target="_blank"
              >
                Learn more
                EuiIconMock
                <span
                  class="euiScreenReaderOnly"
                >
                  (opens in a new tab or window)
                </span>
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      >
        <div
          class="euiFlexItem"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
          >
            <div
              class="euiFlexItem euiFlexItem--flexGrowZero"
            >
              <div
                class="euiPopover euiPopover--anchorDownCenter"
                data-test-subj="moreAction"
              >
                <div
                  class="euiPopover__anchor"
                >
                  <button
                    class="euiButton euiButton--primary euiButton--small"
                    type="button"
                  >
                    <span
                      class="euiButtonContent euiButtonContent--iconRight euiButton__content"
                    >
                      EuiIconMock
                      <span
                        class="euiButton__text"
                      >
                        Actions
                      </span>
                    </span>
                  </button>
                </div>
              </div>
            </div>
            <div
              class="euiFlexItem euiFlexItem--flexGrowZero"
            >
              <button
                class="euiButton euiButton--primary euiButton--small euiButton--fill"
                data-test-subj="Create templateButton"
                type="button"
              >
                <span
                  class="euiButtonContent euiButton__content"
                >
                  EuiIconMock
                  <span
                    class="euiButton__text"
                  >
                    Create template
                  </span>
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <hr
    class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
  />
  <div
    style="padding: initial;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="padding: 0px 5px;"
    >
      <div
        class="euiFlexItem"
      >
        <div
          class="euiFormControlLayout euiFormControlLayout--fullWidth euiFormControlLayout--compressed"
        >
          <div
            class="euiFormControlLayout__childrenWrapper"
          >
            <input
              class="euiFieldSearch euiFieldSearch--fullWidth euiFieldSearch--compressed"
              placeholder="Search..."
              type="search"
              value=""
            />
            <div
              class="euiFormControlLayoutIcons"
            >
              <span
                class="euiFormControlLayoutCustomIcon"
              >
                EuiIconMock
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <hr
      class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
    />
    <div
      class="euiBasicTable euiBasicTable-loading ISM-templates-table"
      data-test-subj="templatesTable"
    >
      <div>
        <div
          class="euiTableHeaderMobile"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsBaseline euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow"
          >
            <div
              class="euiFlexItem euiFlexItem--flexGrowZero"
            >
              <div
                class="euiCheckbox"
              >
                <input
                  aria-label="Select all rows"
                  class="euiCheckbox__input"
                  disabled=""
                  id="_selection_column-checkbox_some_html_id"
                  type="checkbox"
                />
                <div
                  class="euiCheckbox__square"
                />
                <label
                  class="euiCheckbox__label"
                  for="_selection_column-checkbox_some_html_id"
                >
                  Select all rows
                </label>
              </div>
            </div>
            <div
              class="euiFlexItem euiFlexItem--flexGrowZero"
            >
              <div
                class="euiTableSortMobile"
              >
                <div
                  class="euiPopover euiPopover--anchorDownRight"
                >
                  <div
                    class="euiPopover__anchor"
                  >
                    <button
                      class="euiButtonEmpty euiButtonEmpty--primary euiButtonEmpty--xSmall euiButtonEmpty--flushRight"
                      type="button"
                    >
                      <span
                        class="euiButtonContent euiButtonContent--iconRight euiButtonEmpty__content"
                      >
                        EuiIconMock
                        <span
                          class="euiButtonEmpty__text"
                        >
                          Sorting
                        </span>
                      </span>
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <table
          class="euiTable euiTable--responsive"
          id="some_html_id"
          tabindex="-1"
        >
          <caption
            class="euiScreenReaderOnly euiTableCaption"
          />
          <thead>
            <tr>
              <th
                class="euiTableHeaderCellCheckbox"
                scope="col"
              >
                <div
                  class="euiTableCellContent"
                >
                  <div
                    class="euiCheckbox euiCheckbox--inList euiCheckbox--noLabel"
                  >
                    <input
                      aria-label="Select all rows"
                      class="euiCheckbox__input"
                      data-test-subj="checkboxSelectAll"
                      disabled=""
                      id="_selection_column-checkbox_some_html_id"
                      type="checkbox"
                    />
                    <div
                      class="euiCheckbox__square"
                    />
                  </div>
                </div>
              </th>
              <th
                aria-live="polite"
                aria-sort="descending"
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_name_0"
                role="columnheader"
                scope="col"
              >
                <button
                  class="euiTableHeaderButton euiTableHeaderButton-isSorted"
                  data-test-subj="tableHeaderSortButton"
                  type="button"
                >
                  <span
                    class="euiTableCellContent"
                  >
                    <span
                      class="euiTableCellContent__text"
                      title="Template name"
                    >
                      Template name
                    </span>
                    EuiIconMock
                  </span>
                </button>
              </th>
              <th
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_templateType_1"
                role="columnheader"
                scope="col"
              >
                <span
                  class="euiTableCellContent"
                >
                  <span
                    class="euiTableCellContent__text"
                    title="Template type"
                  >
                    Template type
                  </span>
                </span>
              </th>
              <th
                aria-live="polite"
                aria-sort="none"
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_index_patterns_2"
                role="columnheader"
                scope="col"
              >
                <button
                  class="euiTableHeaderButton"
                  data-test-subj="tableHeaderSortButton"
                  type="button"
                >
                  <span
                    class="euiTableCellContent"
                  >
                    <span
                      class="euiTableCellContent__text"
                      title="Index patterns"
                    >
                      Index patterns
                    </span>
                  </span>
                </button>
              </th>
              <th
                aria-live="polite"
                aria-sort="none"
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_order_3"
                role="columnheader"
                scope="col"
              >
                <button
                  class="euiTableHeaderButton"
                  data-test-subj="tableHeaderSortButton"
                  type="button"
                >
                  <span
                    class="euiTableCellContent euiTableCellContent--alignRight"
                  >
                    <span
                      class="euiTableCellContent__text"
                      title="Priority"
                    >
                      Priority
                    </span>
                  </span>
                </button>
              </th>
              <th
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_composed_of_4"
                role="columnheader"
                scope="col"
              >
                <span
                  class="euiTableCellContent euiTableCellContent--alignRight"
                >
                  <span
                    class="euiTableCellContent__text"
                    title="Associated component templates"
                  >
                    Associated component templates
                  </span>
                </span>
              </th>
              <th
                class="euiTableHeaderCell"
                role="columnheader"
                scope="col"
              >
                <span
                  class="euiTableCellContent euiTableCellContent--alignRight"
                >
                  <span
                    class="euiTableCellContent__text"
                    title="Actions"
                  >
                    Actions
                  </span>
                </span>
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              class="euiTableRow"
            >
              <td
                class="euiTableRowCell euiTableRowCell--isMobileFullWidth"
                colspan="7"
              >
                <div
                  class="euiTableCellContent euiTableCellContent--alignCenter"
                >
                  <span
                    class="euiTableCellContent__text"
                  >
                    <div
                      class="euiEmptyPrompt"
                    >
                      <span
                        class="euiTextColor euiTextColor--subdued"
                      >
                        <div
                          class="euiText euiText--medium"
                        >
                          <div
                            class="euiText euiText--small"
                          >
                            <p>
                              You have no templates.
                            </p>
                          </div>
                        </div>
                      </span>
                      <div
                        class="euiSpacer euiSpacer--l"
                      />
                      <div
                        class="euiFlexGroup euiFlexGroup--gutterMedium euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentCenter euiFlexGroup--directionColumn euiFlexGroup--responsive"
                      >
                        <div
                          class="euiFlexItem euiFlexItem--flexGrowZero"
                        >
                          <button
                            class="euiButton euiButton--primary euiButton--small euiButton--fill"
                            type="button"
                          >
                            <span
                              class="euiButtonContent euiButton__content"
                            >
                              EuiIconMock
                              <span
                                class="euiButton__text"
                              >
                                Create template
                              </span>
                            </span>
                          </button>
                        </div>
                      </div>
                    </div>
                  </span>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
